android/src/main/java/flutter/plugins/contactsservice/contactsservice/ContactsServicePlugin.java

```diff
@@ -308,26 +319,62 @@ private void setAvatarDataForContactIfAvailable(Contact contact) {
     }
   }
 
-  private void loadContactPhotoHighRes(Contact contact, boolean photoHighResolution) {
-      try {
-          Uri uri = ContentUris.withAppendedId(ContactsContract.Contacts.CONTENT_URI, Long.parseLong(contact.identifier));
-          InputStream input = ContactsContract.Contacts.openContactPhotoInputStream(contentResolver, uri, photoHighResolution);
-
-          if (input == null){
-              contact.avatar = new byte[0];
-              return;
-          }
+  private void getAvatarForContacts(final List<String> identifiers, final boolean highRes,
+      final Result result) {
+    new GetAvatarsTask(identifiers, highRes, contentResolver, result).execute();
+  }
 
-          Bitmap bitmap = BitmapFactory.decodeStream(input);
-          input.close();
-          ByteArrayOutputStream stream = new ByteArrayOutputStream();
-          bitmap.compress(Bitmap.CompressFormat.PNG, 100, stream);
-          contact.avatar = stream.toByteArray();
-          stream.close();
+  private static class GetAvatarsTask extends AsyncTask<Void, Void, List<byte[]>> {
+    final List<String> identifiers;
+    final boolean highRes;
+    final ContentResolver contentResolver;
+    final Result result;
+
+    GetAvatarsTask(final List<String> identifiers, final boolean highRes,
+        final ContentResolver contentResolver, final Result result) {
+      this.identifiers = identifiers;
+      this.highRes = highRes;
+      this.contentResolver = contentResolver;
+      this.result = result;
+    }
 
-      } catch (IOException e){
-          e.printStackTrace();
+    @Override
+    protected List<byte[]> doInBackground(final Void... params) {
+      // Load avatar for each contact identifier.
+      final List<byte[]> avatars = new ArrayList<>();
+      for (final String identifier : identifiers) {
+        avatars.add(loadContactPhotoHighRes(identifier, highRes, contentResolver));
       }
+
+      return avatars;
+    }
+
+    @Override
+    protected void onPostExecute(final List<byte[]> avatars) {
+      result.success(avatars);
+    }
+  }
+
+  private static byte[] loadContactPhotoHighRes(final String identifier,
+      final boolean photoHighResolution, final ContentResolver contentResolver) {
+    try {
+      Uri uri = ContentUris.withAppendedId(ContactsContract.Contacts.CONTENT_URI, Long.parseLong(identifier));
+      InputStream input = ContactsContract.Contacts.openContactPhotoInputStream(contentResolver, uri, photoHighResolution);
+
+      if (input == null) return null;
+
+      Bitmap bitmap = BitmapFactory.decodeStream(input);
+      input.close();
+
+      ByteArrayOutputStream stream = new ByteArrayOutputStream();
+      bitmap.compress(Bitmap.CompressFormat.PNG, 100, stream);
+      final byte[] bytes = stream.toByteArray();
+      stream.close();
+      return bytes;
+    } catch (IOException e){
+      e.printStackTrace();
```

Comment:
Not a good idea in a contacts plugin. Exception might have pii information.

Use Android Log to print that an exception have occurred.

Reply:
Good point. Really we should propagate this error back to the caller somehow, though this plugin was already printing the stack trace so I just did the same thing here. Didn't think about PII though. Updated to use `io.flutter.Log`, which seems to be the Flutter equivalent of the Android Log utility (let me know if I should be using the regular Android API).

Reply:
You can use the regular Android API.

Reply:
Switched to `android.util.Log`.

Just for my own curiosity, what is the advantage to using the default Android logger over Flutter's logger?

Reply:
It would be consistent with whatever SDK you are using. Imagine your plugin was using FingerprintCompat API. That API would not be using Flutter's logger.

I personally think Flutter's Logger should only be used within the Flutter engine code. Basically they do a lot of `.d()` statements which they only want printed when the app is built with flutter-build-mode == Debug. To plugins and to the libraries they are using, that should not matter.

Reply:
Gotcha, makes sense. Thanks for the clarification.